web3/providers/persistent/subscription_manager.py

```diff
@@ -0,0 +1,274 @@
+import asyncio
+import logging
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Union,
+    cast,
+    overload,
+)
+
+from eth_typing import (
+    HexStr,
+)
+
+from web3.exceptions import (
+    SubscriptionProcessingFinished,
+    TaskNotRunning,
+    Web3TypeError,
+    Web3ValueError,
+)
+from web3.types import (
+    FormattedEthSubscriptionResponse,
+    RPCResponse,
+)
+from web3.utils.subscriptions import (
+    EthSubscription,
+    EthSubscriptionContext,
+)
+
+if TYPE_CHECKING:
+    from web3 import AsyncWeb3  # noqa: F401
+    from web3.providers.persistent import (  # noqa: F401
+        PersistentConnectionProvider,
+        RequestProcessor,
+    )
+
+
+class SubscriptionContainer:
+    def __init__(self) -> None:
+        self.subscriptions: List[EthSubscription[Any]] = []
+        self.subscriptions_by_id: Dict[HexStr, EthSubscription[Any]] = {}
+        self.subscriptions_by_label: Dict[str, EthSubscription[Any]] = {}
+
+    def __len__(self) -> int:
+        return len(self.subscriptions)
+
+    def __iter__(self) -> Iterator[EthSubscription[Any]]:
+        return iter(self.subscriptions)
+
+    def add_subscription(self, subscription: EthSubscription[Any]) -> None:
```

Comment:
Do you think it's worth adding a tighter type here rather than the `Any`? Assuming I'm reading this right, there are well-defined params we allow here, right?

Reply:
If I remember correctly, this allows us to still pass in and handle unknown subscriptions to our library, since we only define the execution-spec'd subscriptions. Alchemy has some for example. I'll see if that was the reason and if so, I think we can leave this be since we define tighter types on the defined subscription classes themselves where we know the results.

Reply:
Cool, I'm good with leaving it!